### DISCUSSION TITLE
How to follow directions using Instrument.measurement(), but also be able to read the last read value without communicating with the instrument?

### DISCUSSION THREAD

**Discussion by SBresler** [pymeasure/pymeasure#687] [Q&A]
Example: 

    ```
    channel_string = Instrument.control(
        "INSTrument?",
        "INSTrument %s",
        """ A string property that controls the selected instrument (channel). This property can be set.
    """,
        validator=strict_discrete_set,
        values=["P6V", "P25V", "N25V"],
    )


is there a way to read the value of channel_string without executing the instrument.control() method?

**Comment by msmttchr** [comment]
I don't think so, you could cache last value written by overriding the `write` method at application level. Could you give some more details about the context?

**Reply by SBresler** [reply]
I have a power supply where you might want to do this:
1. Select the channel out of the list in that snippet. P = positive, N = negative, # is the max voltage of the channel.
2. set the voltage at that channel to a value.

I want to use truncated range to make it so the program checks if the number (voltage) you ask to send is valid - for example 14 is good on P25V but not the others, 0 is good on all, and -3 would be good on N25V only.

should I just query the selected channel as a preamble to the volage setting step?

Trying to write a good, complete (but not overloaded) driver as sort of a pilot test of doing a PR. 

Most of the "fancy" instruments we have have their own python drivers - its our older instruments that get lost in the mix. Those have good labview drivers but I don't want to labview anymore (too hard to do stuff like start a meaningful discussion online about it.)

**Reply by SBresler** [reply]
I suppose one thing that is frustrating me at the moment is that looking between the drivers, they use 3 different strategies:

1. property getter and setter decorators @property and @property.setter. This is explicitly discouraged in the official documentation of pymeasure so I'm trying not to do that. 
2. Instrument.control() which seems to be essentially the same as (1), but abstracted
3. a function definition that uses self.write() or self.ask()

I'm unsure when to use 2 and when to use 3.

**Reply by SBresler** [reply]
```
Extreme5000.voltage = Instrument.control(
    ":VOLT?", ":VOLT %g",
    """ A floating point property that controls the voltage
    in Volts, from -1 to 1 V. This property can be set. """,
    validator=strict_range,
    values=[-1, 1],
    dynamic = True,
)
def set_bipolar_mode(self, enabled = True):
    """Safely switch between bipolar/unipolar mode."""

    # some code to switch off the output first

    if enabled:
        self.mode = "BIPOLAR"
        # set valid range of "voltage" property
        self.voltage_values = [-1, 1]
    else:
        self.mode = "UNIPOLAR"
        # note the "propertyname_parametername" form of the attribute
        self.voltage_values = [0, 1]
```

looking at this snippet from the docs, this doesn't make sense to me because voltage_values isn't defined in the Extreme5000.voltage declaration. so how do the values dynamically change in this example snippet?

I might just be missing something.

**Reply by SBresler** [reply]
From coding_standards.rst:

"Getter and setter functions are discouraged, since properties provide a more fluid experience. Given the extensive tools available for defining properties, detailed in the :ref:`Advanced properties <advanced-properties>` section, these types of properties are preferred."

I see now that it's actually get_voltage() and set_voltage() that are discouraged, so #2 is what you want us to do.

but still, in that very example I posted above, it defines set_bipolar_mode(self, enabled = True).

this is confusing. I'm fine with doing it either way I suppose - I just want to write a lot of these and hammer down the exact way I should be doing it to reduce technical debt down the line.

**Reply by BenediktBurger** [reply]
Python properties are discouraged, they are not prohibited. If `Instrument.control` does not fit your bill, you can use properties.

`voltage_values = [5, 7]` works with some 'magic' behind the lines changing the values argument of voltage.
How exactly it works behind the scenes, I do not know yet, but the documentstion tells you how to use these dynamic properties.
From what you wrote, a dynamic property is, what you want. (see the example of using dynamic properties for different variants of a single instrument).

**Reply by SBresler** [reply]
Yeah, i see that now.

i don’t think there is a way to do the dynamic properties without using functions. Not sure.

If it isn’t possible to have dynamic properties without the addition of setter/getter methods i would think it would make more sense to use methods to make instruments do things - the dynamic properties seems really useful.

thanks!

**Reply by BenediktBurger** [reply]
You can use Python properties for changing attributes of dynamic properties or sometimes a method is the best, especially if you want the device to do something (for example starting a scan, reset etc.).

The idea is, to use properties, whenever it seems natural. If the property can be done with `Instrument.control` (or the setting/measurement variant), use that. If not, use a Python property.

**Reply by SBresler** [reply]
OK, I was writing some more yesterday and I found that for complex commands (some devices allow you to build complex commands).

I think my strategy will be to implement all of the "base level" commands as properties for people who like doing it that way, including the standard documentation for that,

and then for the complext arguments (like select channel 3, set the voltage to this level, and limit the current to these ranges) I will build functions that are a little bit more specific to the device. 

One example is that I would like to have functions that do a lot of things if you tell them to, but do less if you supply less arguments. so instrument.channel_apply(channel = None, voltage = None, voltage_limit = None, current_limit = None)

if you supply nothing, it reads the current active channel, if you specify the channel, it selects the active channel, and then if you supply the next two arguments it supplies those parameters to the selected channel.

The command response time is a little slow at about 200ms for some of these devices, so the compound arguments are useful. Sending longer commands that do more things avoids a lot of dead time.

**Reply by BenediktBurger** [reply]
Sounds like a good plan, as it is a natural approach: You have some individual properties and the possibility of complex configuration settings. You might want to write the comment about the long response time into the device description and hint at the usefulness of compound commands.

**Reply by bilderbuchi** [reply]
> If the property can be done with `Instrument.control` (or the setting/measurement variant), use that. 
> If not, use a Python property.

Yes, this. Extend it with a third layer: Use a method -- especially for "action-/command-like" things like `setup_resistance_measurement()` a method can be a better fit than a property, which still should be attribute-like.

A python property being full methods allows you a bit more flexibility if you need it, but the pymeasure properties are good for I guess ~90% of the interfaces we write.

**Reply by bilderbuchi** [reply]
In your case,
> I want to use truncated range to make it so the program checks if the number (voltage) you ask to send is valid - for example 14 is good on P25V but not the others, 0 is good on all, and -3 would be good on N25V only.
> 
> should I just query the selected channel as a preamble to the volage setting step?

I think the best approach is to implement a python property for the channel selector, and the voltage with a pymeasure control like the `Extreme5000.voltage` above,
In the channel selector setter you adjust the voltage limits accordingly:

```
@property
def channel_selector():
    pass # query channel

@channel_selector.setter
def channel_selector(self, value):
    #...setting channel-related stuff
    # if value == "P25V":
        self.voltage_values = [0, 14]
    # else etc
``` 

Note that you can still use validators in a (property) method to validate the inputs, they are just functions. :-) (there are some instances of this in the code base)
